### ISSUE TITLE
How can I use this in place of the built-in cperl-mode?

### ISSUE THREAD

**Issue by smonff** [HaraldJoerg/cperl-mode#8]
Hi,

This is an help request. I would like to use your `cperl-mode` as soon as possible. I dunno how to override the built-in `cperl-mode`. I use a standard Ubuntu Emacs package (`26.3`) and usually install packages through Melpa.

**Comment by HaraldJoerg** [comment]
Unfortunately, the cperl-mode.el in this repository hasn't been published to any repository from which you can just install it yet.

On the other hand, cperl-mode.el is just _one_ file, so you can just grab it and put it in a directory on your Emacs load-path (~/.emacs.d/site-lisp/ comes into mind).  You can also clone the repository to somewhere and create a symlink in that directory if you want to follow closely, or if you want to switch between branches.

I should explain the branches here:

- The "master" branch contains features which haven't yet been published to the official Emacs repositories, in particular the support for syntax provided by Moo(se) and others.  The diff between this branch and upstream is a mess, though.
 - The "upstream" branch is occasionally synchronized with the official Emacs master branch at https://git.savannah.gnu.org/cgit/emacs.git.  _This is where I spend my time now:_ The work in my master branch will, step by step, be integrated into the upstream version. I'd like to avoid the circles I ran in my own repo, though.  When I'm done with this, "upstream" and "master" should be identical.  
A downside of the "upstream" branch is that it is bundled with Emacs, so occasionally other maintainers will modernize it with stuff coming with Emacs 28.... therefore:
 - The branch "cperl-mode-emacs-26" contains all my bug fixes and enhancements which made their way into upstream, but it is also "downgraded" to work with Emacs 26.

Most likely CPerl mode will at some point lead a double life: bundled with Emacs and bound to the current Emacs version, _and_ released on GNU ELPA, backported to Emacs versions which are available on "stable" distributions.  But before doing such a release, I'd still like to work on a bunch of (old) bugs.

**Comment by smonff** [comment]
Thanks for this explanation. What I wondered is: how do I know your version is loaded and not the built-in one?

**Comment by HaraldJoerg** [comment]
Emacs will load the first version it finds in `load-path`, quite similar to how Perl deals with `@INC`.  To find out which one you got, you can check some of the stuff which isn't available with Emacs 26.3:
 - The command `cperl-set-style` can be used to switch between different indentation styles.  All branches in this repository offer a value "PBP" to follow the style of Damian Conway's book "Perl Best Practices".  If that value isn't available when you call `cperl-set-style`, then you're most probably on vanilla 26.3.
 - The "master" branch here has a value 7.0 for `cperl-version`, while upstream versions have 6.2.

Come to think of it... Given the Zydeco diary you're preparing, I could add Zydeco support to my master branch of cperl-mode.  It comes with a gazillion of new keywords and syntax.  This can't be complete without deeper refactoring, though: cperl-mode doesn't process nested namespace-definitions (produced by keywords like `package` and `class`) correctly.

**Comment by smonff** [comment]
>   The command `cperl-set-style` can be used to switch between different indentation styles.  All branches in this repository offer a value "PBP" to follow the style of Damian Conway's book "Perl Best Practices".  If that value isn't available when you call `cperl-set-style`, then you're most probably on vanilla 26.3.

Oh nice! I looks like I managed to switch on cperl 7.0 then.

**Comment by smonff** [comment]
> Come to think of it... Given the Zydeco diary you're preparing, I could add Zydeco support to my master branch of cperl-mode. It comes with a gazillion of new keywords and syntax. This can't be complete without deeper refactoring, though: cperl-mode doesn't process nested namespace-definitions (produced by keywords like `package` and `class`) correctly.

It would be very pleasant. Nice you eared about the diary, I didn't know it reached so many people... I couldn't really imagine that Zydeco support for `cperl-mode` would maybe be a thing one day. I guess you would have to exchange with Toby Inkster about it. 

[Toby recommended to add some keywords to my editor](https://github.com/tobyink/p5-zydeco/issues/4#issuecomment-689847405), what I somewhat managed to do with the following configuration:

```elisp
(font-lock-add-keywords
 'cperl-mode
 '(("\\(has[[:blank:]]\\|include\\|class\\|role\\|method\\)"
    . font-lock-keyword-face)))
```

Note that I am not an elisp specialist, I just trash some configurations in my `init.el`. It worked so far since 2009 😂

It works well with `font-lock` but the biggest issue on my side is `flycheck`. If I don't disable it, everything looks like:

![image](https://user-images.githubusercontent.com/1219097/93718134-63f7fd80-fb7a-11ea-83a1-346018d53b7d.png)

I cannot blame neither Zydeco nor cperl-mode since I make a rather unofficial use of Zydeco objects through the `include` keyword.

I understand it open doors to a major `cperl-mode` refactoring, and it is difficult to say how Zydeco is going to evolve in the future.

**Comment by HaraldJoerg** [comment]
Ok, so let me snatch this opportunity to show where I'm heading to.

I didn't go for `font-lock-add-keywords`, because CPerl mode is more than just syntax highlighting.  Keywords have different roles in Perl, and there's also the support for navigating a large codebase in CPerl mode.

Let me show (no funky theme, just plain Emacs):
![zydeco pl](https://user-images.githubusercontent.com/32883055/93720010-a7586900-fb86-11ea-9ffd-3c282bee0374.png)
The differences are:

- classes are shown like packages, and methods like subs (which they are, basically).
- imported constants are highlighted

That's no big deal, and also a matter of taste. Class attributes _should_ also be highlighted, but right now `has` doesn't have this magic (I hope I'll be able to do it when Cor happens).  But you'll also note that with vanilla CPerl-mode, the second and all following methods are indented too far unless you fix it manually - which does no longer happen in this repo's CPerl mode (try `indent-region` on the whole buffer).

The next screenshot can not happen with `font-lock-add-keywords` alone:
![zydeco_speedbar](https://user-images.githubusercontent.com/32883055/93720009-a6bfd280-fb86-11ea-8ef8-2b30644f1ac9.png)
This is what `speedbar` gives you for this file (and also for other files in the same directory tree, which I've clipped).  For demonstration, I have unfolded everything, which makes the same methods appear several times.  Whenever I need to navigate in a large codebase written by someone else, I like this stuff.

That screenshot also shows a shortcoming which will be difficult to fix: CPerl mode does not show the namespace hierarchy between the package `What::Ever` and the class `Artist`.  It recognizes the `class` keyword, but it doesn't show the methods as `What::Ever::Artist::create`, as by Zydeco's magic.  It also doesn't distinguish between a class and a package on display, but this is due to my laziness and I guess this could be fixed.

I'll push this Zydeco support within the next days to the master branch of this repository. It needs some cleanup and I'd also like to convert that `zydeco.pl` to a test file for `cperl-mode.el`.

I am at a loss with regard to fixing this for flycheck (or flyspell), because  I don't use these regularly. I know, however, that all tools which rely on PPI or PPR to parse Perl don't understand Zydeco nor any of the other CPAN modules which invent their own syntax, and the authors of these syntax enhancement usually don't bother to provide PPI or PPR support.
